syntax = "proto3";

/*
 * pBFT Protocol Definition
 * Implements the three-phase commit protocol: Pre-prepare, Prepare, Commit
 */

service PBFT {
  // ===== pBFT CORE PROTOCOL =====
  rpc PrePrepare (PrePrepareRequest) returns (PrePrepareResponse);
  rpc Prepare (PrepareRequest) returns (PrepareResponse);
  rpc Commit (CommitRequest) returns (CommitResponse);
  
  // ===== CLIENT API =====
  rpc ClientSubmitBlock (ClientBlockRequest) returns (ClientBlockResponse);
  rpc GetBlockchain (Empty) returns (BlockchainResponse);
  rpc GetNodeStatus (Empty) returns (NodeStatusResponse);
  
  // ===== BYZANTINE FAULT SIMULATION =====
  rpc SetMaliciousBehavior (MaliciousConfigRequest) returns (MaliciousConfigResponse);
}

/* ===== COMMON ===== */
message Empty {}

/* ===== BLOCK STRUCTURE ===== */
message Block {
  int32 block_height = 1;
  string previous_hash = 2;
  string block_hash = 3;
  int64 timestamp = 4;
  string data = 5;  // Optional: transaction data
  int32 view_number = 6;  // pBFT view number
  int32 sequence_number = 7;  // pBFT sequence number
}

/* ===== PRE-PREPARE PHASE ===== */
message PrePrepareRequest {
  int32 view_number = 1;
  int32 sequence_number = 2;
  Block block = 3;
  string primary_id = 4;
  string signature = 5;  // Simplified: just node_id for now
}

message PrePrepareResponse {
  bool accepted = 1;
  string node_id = 2;
  string reason = 3;  // Reason for rejection if any
}

/* ===== PREPARE PHASE ===== */
message PrepareRequest {
  int32 view_number = 1;
  int32 sequence_number = 2;
  string block_hash = 3;
  string node_id = 4;
  string signature = 5;
}

message PrepareResponse {
  bool accepted = 1;
  string node_id = 2;
}

/* ===== COMMIT PHASE ===== */
message CommitRequest {
  int32 view_number = 1;
  int32 sequence_number = 2;
  string block_hash = 3;
  string node_id = 4;
  string signature = 5;
}

message CommitResponse {
  bool accepted = 1;
  string node_id = 2;
}

/* ===== CLIENT API ===== */
message ClientBlockRequest {
  string data = 1;  // Block data/transaction
}

message ClientBlockResponse {
  bool success = 1;
  string message = 2;
  int32 block_height = 3;
}

message BlockchainResponse {
  repeated Block blocks = 1;
  int32 chain_length = 2;
}

message NodeStatusResponse {
  string node_id = 1;
  bool is_primary = 2;
  int32 view_number = 3;
  int32 current_sequence = 4;
  int32 blockchain_height = 5;
  bool is_malicious = 6;
  string malicious_type = 7;
}

/* ===== BYZANTINE FAULT SIMULATION ===== */
message MaliciousConfigRequest {
  bool enable_malicious = 1;
  string malicious_type = 2;  // "silent", "wrong_hash", "double_send", "random"
}

message MaliciousConfigResponse {
  bool success = 1;
  string message = 2;
}
